streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.scheduler.rate
+
+import org.apache.spark.SparkConf
+import org.apache.spark.SparkException
+
+/**
+ * A component that estimates the rate at wich an InputDStream should ingest
+ * elements, based on updates at every batch completion.
+ */
+private[streaming] trait RateEstimator extends Serializable {
+
+  /**
+   * Computes the number of elements the stream attached to this `RateEstimator`
+   * should ingest per second, given an update on the size and completion
+   * times of the latest batch.
+   *
+   * @param time The timetamp of the current batch interval that just finished
+   * @param elements The number of elements that were processed in this batch
+   * @param processingDelay The time in ms that took for the job to complete
+   * @param schedulingDelay The time in ms that the job spent in the scheduling queue
+   */
+  def compute(
+      time: Long,
+      elements: Long,
+      processingDelay: Long,
+      schedulingDelay: Long): Option[Double]
+}
+
+object RateEstimator {
+
+  /**
+   * Return a new RateEstimator based on the value of `spark.streaming.RateEstimator`.
+   *
+   * @return None if there is no configured estimator, otherwise an instance of RateEstimator
+   * @throws IllegalArgumentException if there is a configured RateEstimator that doesn't match any
+   *         known estimators.
+   */
+  def makeEstimator(conf: SparkConf): Option[RateEstimator] =
+    conf.getOption("spark.streaming.RateEstimator") map { estimator =>
```

Comment:
"InputDStream.publish" is not something where its immediately clear from the name on what it means in the context of the InputDstream. In fact, publishing a rate is internal implementation detail of how rate controlling is implemented, and not something that should be fundamentally added as a field to the abstract parent class  InputDStreams. In contrast, the field rateController is something that is more generic and implementation-agnostic and is okay to have as a field in the abstract parent class. 

So I propose this. 
1. Let there be a object method `RateController.isBackpressureEnabled`, which will return true if `spark.streaming.backpressure.enabled`. 
2. Class `ReceiverInputDStream` will create a controller instance of `ReceiverRateController` if the above method returns true. You may or may not wrap that check in an object method, `ReceiverRateController.create(): Option[RateController]`
3. `DirectKafkaRateController` will create a conroller instance of `DirectKafkaRateController` in the same way.